jack.py

```diff
@@ -674,6 +675,38 @@ def transport_stop(self):
         """Stop JACK transport."""
         _lib.jack_transport_stop(self._ptr)
 
+    def set_timebase_callback(self,
+                              conditional, timebase_callback, userdata=None):
+        """Register as timebase master for the JACK subsystem.
+
+        The timebase master registers a callback that updates extended position
+        information such as beats or timecode whenever necessary. Without this
+        extended information, there is no need for this function.
+
+        There is never more than one master at a time. When a new client takes
+        over, the former timebase_callback is no longer called. Taking over the
+        timebase may be done conditionally, so it fails if there was a master
+        already.
+
+        The method may be called whether the client has been activated or not.
+
+        Parameters
+        ----------
+        conditional : int
+            non-zero for a conditional request.
+        timebase_callback : callable
+            realtime function that returns position information.
+        arg :
+            an argument for the timebase_callback function.
+        """
+        @self._callback("JackTimebaseCallback")
+        def callback_wrapper(state, nframes, pos, new_pos, _):
+            return timebase_callback(state, nframes, pos, new_pos, userdata)
+
+        _check(_lib.jack_set_timebase_callback(
+            self._ptr, conditional, callback_wrapper, _ffi.NULL),
+            "Error setting timebase callback")
```

Comment:
Le 27/05/2015 14:08, Matthias Geier a écrit :
>
> In jack.py 
> <https://github.com/spatialaudio/jackclient-python/pull/14#discussion_r31126275>:
>
> > +        Parameters
> > +        ----------
> > +        conditional : int
> > +            non-zero for a conditional request.
> > +        timebase_callback : callable
> > +            realtime function that returns position information.
> > +        arg :
> > +            an argument for the timebase_callback function.
> > +        """
> > +        @self._callback("JackTimebaseCallback")
> > +        def callback_wrapper(state, nframes, pos, new_pos, _):
> > +            return timebase_callback(state, nframes, pos, new_pos, userdata)
> > +
> > +        _check(_lib.jack_set_timebase_callback(
> > +            self._ptr, conditional, callback_wrapper, _ffi.NULL),
> > +            "Error setting timebase callback")
>
> Thanks for the hint about the |errno| module.
> I wasn't aware of that module, but I saw that @amstan 
> <https://github.com/amstan> was also using it, so I cherry-picked his 
> commit (05d6d21 
> <https://github.com/spatialaudio/jackclient-python/commit/05d6d21b232835a605744e54928cd4694038e505>).
>
>     or you don't want to add dependencies?
>
> That's not a problem since that's in the standard library, so it isn't 
> really a dependency.
>
> You can just re-base your changes onto the current master, where the 
> |errno| module is already imported (I just renamed it to |_errno| to 
> not pollute the module namespace).
>
ok
>
>     I will make a dirty "if conditional and -1 ...." ?
>
> There could still be a "normal" error and you wouldn't be able to 
> distinguish between the "normal" error and conditional request with an 
> already active timebase master, right?
>
Yes
>
> I think we can just ignore the JACK bug.
> If anybody actually has a problem with that, they should create a new 
> issue.
>
I check hydrogen and qtractor code, no reference to EBUSY
>
> And probably JACK1 does the right thing?
>
> You will only use |conditional=False|, right?
>
In my app I set conditional to True in order to test : no problem. I 
would like my app to integrate with DAW like ardour or qtractor. But 
those application does not set bbt_offset field so I need to recalculate 
it with frame and bpm. No problem if bpm is constant over the song but 
if tempo change I need bbt_offset and daw set as master timebase. So in 
this case I will not register my app as timebase master and make a 
request with conditional = True.